modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/IdempotentRequestHandler.java

```diff
@@ -0,0 +1,144 @@
+ */
+public class IdempotentRequestHandler {
+
+    private static final Logger log = LogManager.getLogger(IdempotentRequestHandler.class);
+
+    /** @since 11.5 */
+    public static final String HEADER_KEY = "NuxeoIdempotencyKey";
+
+    /** @since 11.5 */
+    public static final String STORE_PROPERTY = "idempotency_keyvaluestore";
+
+    /** @since 11.5 */
+    public static final String TTL_SECONDS_PROPERTY = "idempotency_ttl_seconds";
+
+    /** @since 11.5 */
+    public static final String INPROGRESS_MARKER = "IDEMPOTENCY_INPROGRESS_MARKER";
+
+    /** @since 11.5 */
+    public static final String STATUS_PREFIX = "s_";
+
+    /** @since 11.5 */
+    public static final String CONTENT_PREFIX = "c_";
+
+    protected final String store;
+
+    protected final int ttl;
+
+    public IdempotentRequestHandler(FilterConfig filterConfig) {
+        // default values
+        int ttl = 330;
+        String store = "requestcontroller";
+        // retrieve potential configuration from filter
+        String ttlConfig = filterConfig.getInitParameter(IdempotentRequestHandler.TTL_SECONDS_PROPERTY);
+        if (ttlConfig != null) {
+            try {
+                ttl = Integer.valueOf(ttlConfig);
+            } catch (NumberFormatException e) {
+                log.error(e, e);
+            }
+        }
+        String storeConfig = filterConfig.getInitParameter(IdempotentRequestHandler.STORE_PROPERTY);
+        if (storeConfig != null) {
+            store = storeConfig;
+        }
+        this.store = store;
+        this.ttl = ttl;
+    }
+
+    public IdempotentRequestHandler(String storename, int ttl) {
+        this.store = storename;
+        this.ttl = ttl;
+    }
+
+    public void handle(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        if (!"POST".equals(request.getMethod())) {
+            chain.doFilter(request, response);
+            return;
+        }
+        String key = request.getHeader(IdempotentRequestHandler.HEADER_KEY);
+        if (key == null) {
+            chain.doFilter(request, response);
+            return;
+        }
+        KeyValueStore store = Framework.getService(KeyValueService.class).getKeyValueStore(this.store);
+        String storedContent = store.getString(IdempotentRequestHandler.CONTENT_PREFIX + key);
+        if (storedContent == null) {
+            // new request
+            boolean cleanup = false;
+            store.put(IdempotentRequestHandler.CONTENT_PREFIX + key, IdempotentRequestHandler.INPROGRESS_MARKER, ttl);
+            try {
+                CopyingResponseWrapper wrapper = new CopyingResponseWrapper(response);
+                chain.doFilter(request, wrapper);
+                store.put(IdempotentRequestHandler.CONTENT_PREFIX + key, wrapper.getCaptureAsString(), ttl);
+                store.put(IdempotentRequestHandler.STATUS_PREFIX + key, String.valueOf(wrapper.getStatus()), ttl);
```

Comment:
One of them, probably the content, could have no prefix at all. And for the other I would have used a suffix, to make listings/dumps easier to read.